v1.0.0

LaDracG opened the pull request:
First version.

Notes:
1. For "query" request, when DB missing requested info, if requested days number <= 100, it downloads data from Alpha Vantage in "compact" mode (100 data points); otherwise "full" mode.
2. If data of any date doesn't exist both in local DB and Alpha Vantage, skip that date.
3. If requested stock ticker is invalid, return null.

LaDracG committed:
v1.0.0

First version.

Notes:
1. For "query" request, when DB missing requested info, if requested days number <= 100, it downloads data in Alpha Vantage "compact" mode (100 data points); otherwise "full" mode.
2. If data of any date doesn't exist both in local DB and Alpha Vantage, skip that date.
3. If requested stock ticker is invalid, return null.

GalateaSean commented:
You should get rid of all of the following directories/files: 

- backup (not sure if you want to keep this file, but if you have it in a commit, you don't need to keep a backup around. You will always be able to go back and find it. In general you shouldn't commit commented out code)
- complete
- initial (This looks like the initial greeting controller. If you want to keep this, you can add it to a separate branch and remove all of your new code. Up to you.)
- push-to-pws 
- travis.yml (this is used for Continuous Integration that I don't think you have set up yet so take it out for now)

GalateaSean commented on src/main/java/org/finance_price_service/domain/OneDayPrice.java:
```diff
@@ -0,0 +1,61 @@
+@Entity
+@Table(name = "stock_prices")
+@IdClass(OneDayPriceId.class)
+public class OneDayPrice{
```
use @Value and @Builder lombok annotations on this class:

https://projectlombok.org/features/Builder
https://projectlombok.org/features/Value

@Value will take care of getters. @Builder allows you to create the object like this:

OneDayPrice oneDayPrice = OneDayPrice.builder().symbol().date()... etc... .build();

LaDracG replied:
I checked about @value and it's said that since lombok 1.16.0, constructor provided by @value is private. Since this class is an entity and a public constructor is needed, I need to also use @AllArgsConstructor @NoArgsConstructor. Otherwise, I could use @Getter and @Setter instead of @value.

GalateaSean commented on src/main/java/org/finance_price_service/entrypoint/PricesController.java:
```diff
@@ -0,0 +1,27 @@
+ */
+@RestController
+public class PricesController {
+  @Autowired
```
add a space before and after this bean

GalateaSean commented on src/main/java/org/finance_price_service/entrypoint/PricesController.java:
```diff
@@ -0,0 +1,27 @@
+  @Autowired
+  private QueryLogicService logic;
+  @RequestMapping("/query")
+  public PricesSet response(@RequestParam(value = "symbol") String symbol, @RequestParam(value = "days", defaultValue = "0") int days)
```
add a javadoc to each of these methods.  Because they are the entrypoints for users they are important to have javadocs for. A general rule of thumb is that you should add javadocs for any public method

GalateaSean commented on src/main/java/org/finance_price_service/service/AlphaVantageService.java:
```diff
@@ -0,0 +1,31 @@
+ */
+@Service
+public class AlphaVantageService {
+  @Value("${alphavantage.apikey}")
```
Good job getting the properties working with @Value

GalateaSean commented on src/main/java/org/finance_price_service/service/QueryLogicService.java:
```diff
@@ -0,0 +1,104 @@
+      double low = jsonField.getValue().get(LOW.key).asDouble();
+      double close = jsonField.getValue().get(CLOSE.key).asDouble();
+      int volume = jsonField.getValue().get(VOLUME.key).asInt();
+      price.setValue(symbol, date, open, high, low, close, volume);
```
If you add the @Builder lombok annotation to OneDayPrice, you can create this object with 

OneDayPrice price = OneDayPrice.builder().date(date).open(open).high(high).low(low).close(close).volume(volume).build();

LaDracG replied:
Done, also used @builder for PricesSet and ClosedDate classes

GalateaSean commented on pom.xml:
```diff
@@ -0,0 +1,85 @@
+            <scope>compile</scope>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/mysql/mysql-connector-java -->
+        <dependency>
```
add the lombok dependency

GalateaSean commented:
You don't have to include it in the pull request, but we can work on changing how you parse the JSON from AlphaVantage.  Using JsonNode works, but because the AlphaVantage format is unchanging, it would be neater if you create custom classes to deserialize into. I'm going to raise a new issue for this and we can add it in after this gets merged. Added issue #3

GalateaSean commented on src/main/java/org/finance_price_service/service/QueryLogicService.java:
```diff
@@ -0,0 +1,104 @@
+@Service
+public class QueryLogicService {
+  @Autowired
+  private MySQLService sql;
```
rename these beans to be a bit more descriptive -- mysqlService and alphaVantageService  seem like they would be appropriate names. If those seem too long to you, sqlService and alphaService would also be fine

GalateaSean commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,18 @@
+spring:
```
good job with this yml file 👍

GalateaSean commented on test/run.sh:
```diff
@@ -0,0 +1,37 @@
+cd $(dirname $0)
```
I'm not really sure what this file is meant for. We can talk about it tomorrow.

GalateaSean commented on src/main/java/org/finance_price_service/service/QueryLogicService.java:
```diff
@@ -0,0 +1,104 @@
+
+    /**
+     * Converts JsonNode to PricesSet
+     */
```
javadocs ( /** ... **/ )  are usually reserved for method and class descriptions.  If you want to add a comment in the middle of a method like this,  use two slashes like this:  //

GalateaSean commented on pom.xml:
```diff
@@ -0,0 +1,85 @@
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.springframework</groupId>
```
These three tags are used by maven to create and name your jar when you make a build.

https://maven.apache.org/guides/mini/guide-naming-conventions.html

You want your groupId to match with your package name in src/main/java. So you could do something like org.galatea.jingyang as your group id and change your package structure to src/main/java/org/galatea/jingyang

GalateaSean commented on pom.xml:
```diff
@@ -0,0 +1,85 @@
+
+    <groupId>org.springframework</groupId>
+    <artifactId>financial-price-service</artifactId>
+    <version>0.1.0</version>
```
We briefly went over versioning, but this is where it is set for your jar. I didn't mention that the suffix "-SNAPSHOT" is used when you are making a jar that you do not plan to release to production, these jars are just used for testing and development. So set the version to "0.0.0-SNAPSHOT" for now.

LaDracG committed:
Remove unnecessary files

LaDracG committed:
Used Lombok annotations, and other code optimizations

1. Used Lombok annotations to replace manual constructors and other parts.
2. Used more descriptive variable names.
3. Added missing javadocs for some public methods.
4. Added spaces for cleaner code style.

LaDracG committed:
Changed package structure, groupID and version.

LaDracG committed:
Modified codes according to SonarLint suggestions

LaDracG committed:
Revert "Modified codes according to SonarLint suggestions"

This reverts commit 7ea6038f76c100d484941f6ca2a7f9d99d3790f0.

LaDracG committed:
Redesigned data querying strategy (JavaDoc to be updated)

1. Now only executes one MySQL for a request
2. Now only updates until requested date when data missing in DB
3. Updating process skips those existing records, instead of re-inserting them
4. Removed "update" request function (was used only for testing)
5. Code style improvements

LaDracG committed:
Reformed AlphaVantageAPIWords as utility class

LaDracG committed:
Added Javadoc
